mlrun/api/crud/function_marketplace.py

```diff
@@ -0,0 +1,179 @@
+        self._catalogs = {}
+
+    @staticmethod
+    def _get_k8s():
+        k8s_helper = get_k8s()
+        if not k8s_helper.is_running_inside_kubernetes_cluster():
+            return None
+        return k8s_helper
+
+    def add_source(self, source: MarketplaceSource):
+        source_name = source.metadata.name
+        credentials = source.spec.credentials
+        if credentials:
+            self._store_source_credentials(source_name, credentials)
+
+    def remove_source(self, source_name):
+        self._catalogs.pop(source_name, None)
+        if not self._get_k8s():
+            return
+
+        secret_prefix = source_name + secret_name_separator
+        source_credentials = self._get_source_credentials(source_name)
+        if not source_credentials:
+            return
+        secrets_to_delete = [secret_prefix + key for key in source_credentials]
+        self._get_k8s().delete_project_secrets(
+            self._internal_project_name, secrets_to_delete
+        )
+
+    def _store_source_credentials(self, source_name, credentials: dict):
+        if not self._get_k8s():
+            raise mlrun.errors.MLRunInvalidArgumentError(
+                "MLRun is not configured with k8s, marketplace source credentials cannot be stored securely"
+            )
+        secret_prefix = source_name + secret_name_separator
+
+        adjusted_credentials = {
+            secret_prefix + key: value for key, value in credentials.items()
+        }
+        self._get_k8s().store_project_secrets(
+            self._internal_project_name, adjusted_credentials
+        )
+
+    def _get_source_credentials(self, source_name):
+        if not self._get_k8s():
+            return {}
+
+        secret_prefix = source_name + secret_name_separator
+        secrets = self._get_k8s().get_project_secret_values(self._internal_project_name)
+        source_secrets = {}
+        for key, value in secrets.items():
+            if key.startswith(secret_prefix):
+                source_secrets[key[len(secret_prefix) :]] = value
+
+        return source_secrets
+
+    @staticmethod
+    def _transform_catalog_dict_to_schema(source, catalog_dict):
+        catalog_dict = catalog_dict.get("functions")
+        if not catalog_dict:
+            raise mlrun.errors.MLRunInternalServerError(
+                "Invalid catalog file - no 'functions' section found."
+            )
+
+        catalog = MarketplaceCatalog(catalog=[])
+        # Loop over channels, then per function extract versions.
+        for channel_name in catalog_dict:
+            channel_dict = catalog_dict[channel_name]
+            for function_name in channel_dict:
+                function_dict = channel_dict[function_name]
+                for version_tag in function_dict:
+                    version_dict = function_dict[version_tag]
+                    metadata_dict = version_dict.copy()
+                    spec_dict = metadata_dict.pop("spec", None)
```

Comment:
I'll rename the `metadata_dict` to something else. It's basically a dump of all function details, and currently most of the fields are in the root of the dict, and there's a sub-dictionary called `spec`. This should change in the future but for now it is what it is. 
Since this object is created by another tool, I have not modified the content it generates, I'm just adjusting to it.